### PR TITLE
conditionals: now with topological sort

### PR DESCRIPTION
**Pull request by tom-wa**
# Purpose

Added topological sorting 
# Checklist
- [x] commit messages are fine (with references to issues)
- [x] I ran all tests and everything went fine
- [ ] I added unit tests
- [ ] affected documentation is fixed
- [ ] I added code comments, logging, and assertions
- [ ] meta data is updated (e.g. README.md of plugins)

### CODE REVIEW

File: src/plugins/conditionals/conditionals.c [hunk 1]
```diff
@@ -785,6 +787,129 @@ static CondResult evaluateKey (const Key * meta, const Key * suffixList, Key * p
 	return TRUE;
 }
 
+static KeySet * getDependencies (const char * expr, Key * cur, Key * parentKey)
+{
+	char * localExpr = strdup (expr);
```
Comment (markus2330) [inline src/plugins/conditionals/conditionals.c]: please use elektraStrDup.

File: src/plugins/conditionals/conditionals.c [hunk 2]
```diff
@@ -785,6 +787,129 @@ static CondResult evaluateKey (const Key * meta, const Key * suffixList, Key * p
...
+	char * ptr = localExpr;
+	KeySet * deps = ksNew (0, KS_END);
+	int isEscaped = 0;
+	int hasQuote = 0;
+	for (; *ptr; ++ptr)
```
Comment (markus2330) [inline src/plugins/conditionals/conditionals.c]: Shouldn't this parsing code be part of libmeta (later)? Some generic link detection would be nice.

**Comment by markus2330** [comment]
Please add unit tests. What else is missing?

**Comment by tom-wa** [comment]
>Shouldn't this parsing code be part of libmeta (later)?

you are right. currently conditionals is the only plugin that benefits from it though, so let's keep it that way for now, i'll improve it and move it to libmeta later.

>What else is missing?

a lot of unit tests, +a few things to discuss: 
for now, should the topological sort be used by default or activated by a flag (until it's proven to work) ?
what should be done if there are cyclic dependencies or invalid dependencies ?

**Comment by markus2330** [comment]
> should the topological sort be used by default

I think we should not do this for the next release anyway, but the idea is that it should be default when any calculations are involved.

In longer term we should split the plugin to a validation and a calculation plugin. The calculation should be done at lookup time (see #867). Only the calculation needs correct ordering, for validation the order only matters for better error messages, but is not needed for correctness.

**Comment by markus2330** [comment]
I'll close it. Maybe we can continue this after Elektra 1.0.

**Pull request closed without merging**